### ISSUE TITLE
asm! macro does not mark variable as used

### ISSUE THREAD

**Issue by Soveu** [rust-lang/rust#77915]
I tried this code:

```rust
unsafe fn rep_movsb(mut dest: *mut u8, mut src: *const u8, mut n: usize) -> *mut u8 {
    while n != 0 {
        asm!(
            "rep movsb",
            inout("rcx") n,
            inout("rsi") src,
            inout("rdi") dest,
        );
    }

    return dest;
}
```

I expected to see this happen:
Compiler doesn't generate any warnings

Instead, this happened:
Rustc says that `src` is unused, even though it is declared as `inout` in asm.
Adding arrows (`inout("rsi") src => src`) fixes it

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (8dae8cdcc 2020-10-12)
binary: rustc
commit-hash: 8dae8cdcc8fa879cea6a4bbbfa5b32e97be4c306
commit-date: 2020-10-12
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
LLVM version: 11.0
```

**Comment by oliviacrain** [comment]
Interesting bug! I used `cargo bisect-rustc` to determine that this bug has been around [since the commit introducing the `asm!` syntax](https://github.com/rust-lang/rust/commit/6cb1f5972dffd7614199a828ecba140606e6e65e). 

[Looking at the MCVE case from above](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ce6747a4ea6c75a03976813e77bb6c89) and doing some debugging, I found that for all live nodes in `rep_movsb` , `src` is never marked as being used in the RWU table. It only gets marked as read/written. [Later on](https://github.com/rust-lang/rust/blob/f42692b1cc1552abf905f9b6650061d246024252/compiler/rustc_passes/src/liveness.rs#L1615-L1617), this is part of the information used to determine that `src` is unused. That doesn't really make much sense though- [the definition of a use](https://github.com/rust-lang/rust/blob/f42692b1cc1552abf905f9b6650061d246024252/compiler/rustc_passes/src/liveness.rs#L74-L77) does seem to fit the case where a variable is used as both an input and an output to some arbitrary assembly instruction.

I'll go ahead and claim this. I have a solution in mind, just going to ask around zulip first before putting out a PR.

@rustbot claim
